File: puzzles/js/roller-coaster.js
let inputs = readline().split(' ');
const NB_PLACES = parseInt(inputs[0]);
const NB_RIDES = parseInt(inputs[1]);
const NB_GROUPS = parseInt(inputs[2]);

let groups = [];
for (let i = 0; i < NB_GROUPS; ++i) {
  let nbPeople = parseInt(readline());
  groups.push(nbPeople);
}

// Calculate earnings starting from each group (1 ride)
// Save the result (dynamic programming)
let results = [];
for (let i = 0; i < NB_GROUPS; ++i) {
  let index = i;
  let nbPeople = groups[index];
  let nbPeopleRiding = 0;
  while (nbPeopleRiding + nbPeople <= NB_PLACES) {
    nbPeopleRiding += nbPeople;
    if (index < NB_GROUPS - 1) {
      index += 1
    } else {
      index = 0
    }
    // all groups are riding already?
    if (index == i) {
      break;
    }
    nbPeople = groups[index];
  }
  results.push({
    "nbPeopleRiding": nbPeopleRiding,
    "nextGroup" : index
  });
}

// Calculate the total earnings starting from the first group (all rides)
let totalMoney = 0;
let currentGroup = 0;
for (let i = 0; i < NB_RIDES; ++i) {
  let result = results[currentGroup];
  totalMoney += result.nbPeopleRiding;
  currentGroup = result.nextGroup;
}

print(totalMoney)

File: puzzles/cpp/roller-coaster.cpp
#include <iostream>
#include <vector>
#include <utility>

using namespace std;

int main() {
  int nbPlaces, nbRides, nbGroups;
  cin >> nbPlaces >> nbRides >> nbGroups; cin.ignore();
  vector<long> groups;
  for (int i = 0; i < nbGroups; i++) {
    long nbPeople;
    cin >> nbPeople; cin.ignore();
    groups.push_back(nbPeople);
  }

  // Calculate earnings starting from each group (1 ride)
  // Save the result (dynamic programming)
  vector<pair<long, int>> results;
  for (int i = 0; i < nbGroups; i++) {
    int index = i;
    long nbPeople = groups[index];
    long nbPeopleRiding = 0;
    while (nbPeopleRiding + nbPeople <= nbPlaces) {
      nbPeopleRiding += nbPeople;
      if (index < nbGroups - 1) {
        index++;
      } else {
        index = 0;
      }
      // all groups are riding already?
      if (index == i) {
        break;
      }
      nbPeople = groups[index];
    }
    results.push_back(make_pair(nbPeopleRiding, index));
  }

  // Calculate the total earnings starting from the first group (all rides)
  long totalMoney = 0;
  int currentGroup = 0;
  for (int i = 0; i < nbRides; i++) {
    totalMoney += results[currentGroup].first;
    currentGroup = results[currentGroup].second;
  }
  cout << totalMoney << endl;
}
